lib/mix/lib/mix/compilers/elixir.ex

```diff
@@ -206,44 +234,62 @@ defmodule Mix.Compilers.Elixir do
   # files that have changed. It then, recursively, figures out
   # all the files that changed (via the module dependencies) and
   # return the non-changed entries and the removed sources.
-  defp update_stale_entries(all, [], stale) when stale == %{} do
-    {all, []}
+  defp update_stale_entries(modules, files, [], stale) when stale == %{} do
+    {modules, files, []}
   end
 
-  defp update_stale_entries(all, changed, stale) do
+  defp update_stale_entries(modules, files, changed, stale) do
     removed = Enum.into(changed, %{}, &{&1, true})
-    remove_stale_entries(all, stale, removed)
+    remove_stale_entries(modules, files, stale, removed)
   end
 
-  defp remove_stale_entries(entries, old_stale, old_removed) do
-    {rest, new_stale, new_removed} =
-      Enum.reduce entries, {[], old_stale, old_removed}, &remove_stale_entry/2
+  defp remove_stale_entries(modules, files, old_stale, old_removed) do
+    file_modules =
+      Enum.reduce modules, %{}, fn module(source: source) = module, file_modules ->
+        file = Enum.find(files, & file(&1, :file) == source)
+        Map.update(file_modules, file, [module], &[module | &1])
+      end
+
+    {rest_modules, rest_files, new_stale, new_removed} =
+      Enum.reduce file_modules, {[], [], old_stale, old_removed}, &remove_stale_entry/2
 
     if map_size(new_stale) > map_size(old_stale) or
        map_size(new_removed) > map_size(old_removed) do
-      remove_stale_entries(rest, new_stale, new_removed)
+      remove_stale_entries(rest_modules, rest_files, new_stale, new_removed)
     else
-      {rest, Map.keys(new_removed)}
+      {rest_modules, rest_files, Map.keys(new_removed)}
     end
   end
 
-  defp remove_stale_entry({beam, module, _kind, source, compile, runtime, _bin} = entry,
-                          {rest, stale, removed}) do
+  defp remove_stale_entry({file, modules}, acc) do
+    file(file: source, compile: compile, runtime: runtime) = file
+    {rest_modules, rest_files, stale, removed} = acc
+
     cond do
       # If I changed in disk or have a compile time dependency
       # on something stale, I need to be recompiled.
       Map.has_key?(removed, source) or Enum.any?(compile, &Map.has_key?(stale, &1)) ->
-        remove_and_purge(beam, module)
-        {rest, Map.put(stale, module, true), Map.put(removed, source, true)}
+        stale =
+          Enum.reduce modules, stale, fn module(beam: beam, module: module), stale ->
+            remove_and_purge(beam, module)
+
```

Comment:
@whatyouhide ah sorry, just my personal style again. I like to split code up into "paragraphy" kinda things.